File: java/src/main/java/com/purescraps/purehtml/transformers/types/JSONPathTransformer.java
package com.purescraps.purehtml.transformers.types;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import com.purescraps.purehtml.transformers.TransformParams;
import com.purescraps.purehtml.transformers.Transformer;

/**
 * jsonpath: Extract a value from a parsed JSON value by path
 *
 * Usage:
 * jsonpath(path): navigates a JSON value (typically the output of `json`)
 * using dot notation for object keys (`$.a.b`) and bracket notation for
 * array indices or keys with special characters (`$.a[0]`, `$["a-b"]`).
 * Negative indices count from the end of the array. Returns null when the
 * path does not resolve to a value.
 */
public class JSONPathTransformer extends Transformer {
    private List<String> args;

    public JSONPathTransformer(List<String> args) {
        this.args = args;
    }

    public JSONPathTransformer() {
        this.args = null;
    }

    public void setArgs(List<String> args) {
        this.args = args;
    }

    public String getName() {
        return "jsonpath";
    }

    @Override
    public Object transform(TransformParams params) {
        Object val = params.getVal();

        if (val == null) {
            return null;
        }

        if (args == null || args.isEmpty()) {
            throw new IllegalArgumentException("jsonpath: requires a path, e.g. jsonpath(\"$.a.b\")");
        }

        List<Segment> segments = parsePath(args.get(0));
        return evaluate(val, segments);
    }

    private abstract static class Segment {
    }

    private static final class KeySegment extends Segment {
        final String key;

        KeySegment(String key) {
            this.key = key;
        }
    }

    private static final class IndexSegment extends Segment {
        final int index;

        IndexSegment(int index) {
            this.index = index;
        }
    }

    private static boolean isIdentChar(char c) {
        return Character.isLetterOrDigit(c) || c == '_';
    }

    private static List<Segment> parsePath(String path) {
        if (path == null || path.isEmpty() || path.charAt(0) != '$') {
            throw new IllegalArgumentException("jsonpath: invalid path \"" + path + "\": must start with \"$\"");
        }

        List<Segment> segments = new ArrayList<>();
        int i = 1;
        int n = path.length();

        while (i < n) {
            char ch = path.charAt(i);
            if (ch == '.') {
                i++;
                int start = i;
                while (i < n && isIdentChar(path.charAt(i))) {
                    i++;
                }
                if (i == start) {
                    throw new IllegalArgumentException(
                            "jsonpath: invalid path \"" + path + "\": expected a name after \".\"");
                }
                segments.add(new KeySegment(path.substring(start, i)));
            } else if (ch == '[') {
                i++;
                if (i >= n) {
                    throw new IllegalArgumentException(
                            "jsonpath: invalid path \"" + path + "\": unterminated \"[\"");
                }
                char next = path.charAt(i);
                if (next == '"' || next == '\'') {
                    char quote = next;
                    i++;
                    StringBuilder key = new StringBuilder();
                    while (i < n && path.charAt(i) != quote) {
                        char c = path.charAt(i);
                        char nextChar = (i + 1 < n) ? path.charAt(i + 1) : '\0';
                        if (c == '\\' && i + 1 < n && (nextChar == quote || nextChar == '\\')) {
                            key.append(nextChar);
                            i += 2;
                        } else {
                            key.append(c);
                            i++;
                        }
                    }
                    if (i >= n) {
                        throw new IllegalArgumentException(
                                "jsonpath: invalid path \"" + path + "\": unterminated quoted key");
                    }
                    i++; // skip closing quote
                    segments.add(new KeySegment(key.toString()));
                } else if (next == '-' || Character.isDigit(next)) {
                    int start = i;
                    if (next == '-') {
                        i++;
                    }
                    int digitsStart = i;
                    while (i < n && Character.isDigit(path.charAt(i))) {
                        i++;
                    }
                    if (i == digitsStart) {
                        throw new IllegalArgumentException(
                                "jsonpath: invalid path \"" + path + "\": expected an index inside \"[]\"");
                    }
                    segments.add(new IndexSegment(Integer.parseInt(path.substring(start, i))));
                } else {
                    throw new IllegalArgumentException(
                            "jsonpath: invalid path \"" + path + "\": expected an index or quoted key inside \"[]\"");
                }
                if (i >= n || path.charAt(i) != ']') {
                    throw new IllegalArgumentException(
                            "jsonpath: invalid path \"" + path + "\": expected \"]\"");
                }
                i++;
            } else {
                throw new IllegalArgumentException(
                        "jsonpath: invalid path \"" + path + "\": unexpected character '" + ch + "' at position " + i);
            }
        }

        return segments;
    }

    @SuppressWarnings("unchecked")
    private static Object evaluate(Object root, List<Segment> segments) {
        Object current = root;

        for (Segment segment : segments) {
            if (current == null) {
                return null;
            }

            if (segment instanceof IndexSegment) {
                if (!(current instanceof List)) {
                    return null;
                }
                List<Object> list = (List<Object>) current;
                int index = ((IndexSegment) segment).index;
                if (index < 0) {
                    index += list.size();
                }
                if (index < 0 || index >= list.size()) {
                    return null;
                }
                current = list.get(index);
            } else {
                if (!(current instanceof Map)) {
                    return null;
                }
                Map<Object, Object> map = (Map<Object, Object>) current;
                String key = ((KeySegment) segment).key;
                if (!map.containsKey(key)) {
                    return null;
                }
                current = map.get(key);
            }
        }

        return current;
    }
}

File: python/purehtml/transformers/types/jsonpath_transformer.py
from abc import ABC
from typing import List, Union

from purehtml.transformers.transform_params import TransformParams
from purehtml.transformers.transformer import Transformer


class _KeySegment:
    __slots__ = ("key",)

    def __init__(self, key: str):
        self.key = key


class _IndexSegment:
    __slots__ = ("index",)

    def __init__(self, index: int):
        self.index = index


_Segment = Union[_KeySegment, _IndexSegment]


def _is_ident_char(ch: str) -> bool:
    return ch.isalnum() or ch == "_"


def _parse_json_path(path: str) -> List[_Segment]:
    """
    Parses a minimal jsonpath subset: dot notation for object keys
    (`$.a.b`), bracket notation with a quoted key for keys with special
    characters (`$["a-b"]`), and bracket notation with an integer for array
    indices, including negative indices counted from the end (`$.a[-1]`).
    """
    if not path or path[0] != "$":
        raise ValueError(f'jsonpath: invalid path {path!r}: must start with "$"')

    segments: List[_Segment] = []
    i = 1
    n = len(path)

    while i < n:
        ch = path[i]
        if ch == ".":
            i += 1
            start = i
            while i < n and _is_ident_char(path[i]):
                i += 1
            if i == start:
                raise ValueError(f'jsonpath: invalid path {path!r}: expected a name after "."')
            segments.append(_KeySegment(path[start:i]))
        elif ch == "[":
            i += 1
            if i >= n:
                raise ValueError(f'jsonpath: invalid path {path!r}: unterminated "["')
            if path[i] in ("\"", "'"):
                quote = path[i]
                i += 1
                chars = []
                while i < n and path[i] != quote:
                    if path[i] == "\\" and i + 1 < n and path[i + 1] in (quote, "\\"):
                        chars.append(path[i + 1])
                        i += 2
                    else:
                        chars.append(path[i])
                        i += 1
                if i >= n:
                    raise ValueError(f"jsonpath: invalid path {path!r}: unterminated quoted key")
                i += 1  # skip closing quote
                segments.append(_KeySegment("".join(chars)))
            elif path[i] == "-" or path[i].isdigit():
                start = i
                if path[i] == "-":
                    i += 1
                digits_start = i
                while i < n and path[i].isdigit():
                    i += 1
                if i == digits_start:
                    raise ValueError(f'jsonpath: invalid path {path!r}: expected an index inside "[]"')
                segments.append(_IndexSegment(int(path[start:i])))
            else:
                raise ValueError(f'jsonpath: invalid path {path!r}: expected an index or quoted key inside "[]"')
            if i >= n or path[i] != "]":
                raise ValueError(f'jsonpath: invalid path {path!r}: expected "]"')
            i += 1
        else:
            raise ValueError(f'jsonpath: invalid path {path!r}: unexpected character {ch!r} at position {i}')

    return segments


def _evaluate_json_path(root, segments: List[_Segment]):
    current = root

    for segment in segments:
        if current is None:
            return None

        if isinstance(segment, _IndexSegment):
            if not isinstance(current, list):
                return None
            index = segment.index
            if index < 0:
                index += len(current)
            if index < 0 or index >= len(current):
                return None
            current = current[index]
        else:
            if not isinstance(current, dict):
                return None
            if segment.key not in current:
                return None
            current = current[segment.key]

    return current


class JSONPathTransformer(Transformer, ABC):
    """
    jsonpath: Extract a value from a parsed JSON value by path

    Usage:
    jsonpath(path): navigates a JSON value (typically the output of `json`)
    using dot notation for object keys (`$.a.b`) and bracket notation for
    array indices or keys with special characters (`$.a[0]`, `$["a-b"]`).
    Negative indices count from the end of the array. Returns None when the
    path does not resolve to a value.
    """

    def __init__(self, args: List[str] = None):
        super().__init__()
        self.args = args if args is not None else []

    def set_args(self, args: List[str]):
        self.args = args

    def get_name(self) -> str:
        return "jsonpath"

    def transform(self, params: TransformParams):
        val = params.get_val()

        if val is None:
            return None

        if len(self.args) < 1:
            raise ValueError('jsonpath: requires a path, e.g. jsonpath("$.a.b")')

        segments = _parse_json_path(self.args[0])
        return _evaluate_json_path(val, segments)
